no-jira: add retry_on_exception

ColinDKelley opened the pull request:
## [0.4.0] - Unreleased
### Added
- Added `Invoca::Utils.retry_on_exception`.  (Copied from `web` repo. Will remove there later.)
- Added `Invoca::Utils::GuaranteedUTF8String.normalize_all_strings` to normalize
  all strings found in a JSON doc of hashes, arrays, and values.

ColinDKelley committed:
no-jira: add retry_on_exception

ColinDKelley committed:
no-jira: require exceptions

ColinDKelley committed:
no-jira: require rake_test_warning_false

ColinDKelley committed:
no-jira: bundle update; add test_overrides

ColinDKelley committed:
no-jira: bump version to 0.4.0 and update CHANGELOG

ColinDKelley committed:
no-jira: omit redundant begin/end; update inline rdocs

jebentier commented on lib/invoca/utils/exceptions.rb:
```diff
@@ -0,0 +1,25 @@
+      # @param [Integer] retries: - 1+ count of retries (1 retry = up to 2 tries total)
+      # @param [Proc] before_retry - optional proc which is called before each retry, with the exception passed as a block param
+      # @return the value from yield
+      def retry_on_exception(exception_classes, retries:, before_retry: nil)
```
I know this probably is just a direct copy over, but I have a couple questions for the new definition.

1. should we default `retries` to 1 as a sane default?
2. should `exception_classes` default to to a catch all `StandardError`?
3. if the answer to 1 and 2 are no, would it be better to make all the arguments kwargs?

ColinDKelley replied:
> 1. should we default retries to 1 as a sane default?

I often don't go for defaults when there are no callers--on the reasoning that explicit is better--but I suppose it would be pretty clear that `retry_on_exception` retries just once.

> 2. should exception_classes default to to a catch all StandardError?

Nah, in all the case I've used this method or a similar one, I've never used `StandardError`. It's always been more targeted than that.

I did consider making the exception classes argument `*exception_classes` so you could provide a list without `[` `]` around them. But I didn't do that, because it's pretty common for the exception list to be in a big constant, and if I use use `*` to scoop them up, that forces the caller to use `*` to splat them out.

ColinDKelley committed:
no-jira: default retries: to 1

ColinDKelley committed:
no-jira: add normalize_all_strings

ColinDKelley committed:
no-jira: update CHANGELOG

ColinDKelley commented:
@jebentier I added the default of `retries: 1` and a test. Also, I added `Invoca::Utils::GuaranteedUTF8String.normalize_all_strings` so I can use it in TECH-4425 to normalize the strings in an entire logging hash in one pass.

ColinDKelley committed:
no-jira: release v0.4.0
